File: samples/typescript/src/client/interactive-handler.ts
/**
 * Interactive handler for CMS SDK
 * CMS SDK 交互处理器
 */

import * as readline from 'readline';
import { EventType, InteractionType } from '../types/index.js';
import { AgentClient, ChatEvent } from './agent-client.js';
import { SDKException, ErrorCode } from './errors.js';

/** 交互响应 / Interactive response */
export interface InteractiveResponse {
  callId: string;
  type: InteractionType;
  response: Record<string, unknown>;
  source?: Record<string, unknown>;
  modifiedData?: Record<string, unknown>;
  formData?: Record<string, unknown>;
  decision?: string;
}

/** 交互事件处理器 / Interactive event handler */
export class InteractiveHandler {
  private client: AgentClient;
  private timeout?: number;
  private rl?: readline.Interface;

  constructor(client: AgentClient, timeout?: number) {
    this.client = client;
    this.timeout = timeout;
  }

  /** 处理交互事件 / Handle interactive event */
  async handleEvent(event: Record<string, unknown>, callId: string): Promise<InteractiveResponse> {
    if (!event) {
      throw new SDKException(ErrorCode.PARSE_ERROR, '事件为空');
    }

    const eventType = event.type as string;
    if (eventType !== EventType.INTERACTIVE) {
      throw new SDKException(ErrorCode.PARSE_ERROR, `不支持的事件类型: ${eventType}`);
    }

    const payload = event.payload as Record<string, unknown>;
    if (!payload) {
      throw new SDKException(ErrorCode.PARSE_ERROR, '交互负载为空');
    }

    const interactiveType = payload.type as string;

    switch (interactiveType) {
      case InteractionType.USER_ACK:
        return this.handleUserAck(payload, callId);
      case InteractionType.USER_SELECT:
        return this.handleUserSelect(payload, callId);
      case InteractionType.USER_INPUT:
        return this.handleUserInput(payload, callId);
      default:
        throw new SDKException(ErrorCode.PARSE_ERROR, `不支持的交互类型: ${interactiveType}`);
    }
  }

  /** 处理用户确认 / Handle user acknowledgment */
  async handleUserAck(payload: Record<string, unknown>, callId: string): Promise<InteractiveResponse> {
    const title = this.getTitle(payload);
    const message = this.getDescription(payload);
    const options = this.getOptions(payload);
    const modifiedData = this.extractData(payload);

    console.log('\n🔔 确认请求');
    console.log('--------------');
    if (title) console.log(title);
    if (message) console.log(`\n${message}`);

    if (modifiedData) {
      console.log();
      for (const [key, value] of Object.entries(modifiedData)) {
        if (key === 'title' || key === 'message') continue;
        console.log(`${key}: ${value}`);
      }
    }
    console.log('--------------');

    if (options.length > 0) {
      const parts = options.map((opt, i) =>
        `[${this.getOptionValue(opt)}] ${this.getOptionLabel(opt, i)}`);
      await this.prompt(`请输入 ${parts.join(', ')}: `);
    } else {
      await this.prompt('请输入 [y/yes] 确认，[n/no] 取消: ');
    }

    const input = (await this.prompt('')).trim().toLowerCase();
    // TODO: the above is wrong — we need to capture the prompt's result
    // Actually, let me restructure this properly

    let userInput: string;
    if (options.length > 0) {
      const parts = options.map((opt, i) =>
        `[${this.getOptionValue(opt)}] ${this.getOptionLabel(opt, i)}`);
      userInput = await this.prompt(`请输入 ${parts.join(', ')}: `);
    } else {
      userInput = await this.prompt('请输入 [y/yes] 确认，[n/no] 取消: ');
    }

    userInput = userInput.trim().toLowerCase();
    let confirmed = ['', 'y', 'yes', '是'].includes(userInput);
    let decision = confirmed ? 'yes' : 'no';

    for (const opt of options) {
      if (userInput === this.getOptionValue(opt).toLowerCase()) {
        decision = this.getOptionValue(opt);
        confirmed = true;
        break;
      }
    }

    return {
      callId,
      type: InteractionType.USER_ACK,
      response: { confirmed },
      source: this.extractSource(payload),
      modifiedData,
      decision,
    };
  }

  /** 处理用户选择 / Handle user selection */
  async handleUserSelect(payload: Record<string, unknown>, callId: string): Promise<InteractiveResponse> {
    const title = this.getTitle(payload);

    console.log('\n📋 请选择');
    if (title) console.log(`   标题: ${title}`);

    const options = this.getOptions(payload);
    if (options.length === 0) {
      throw new SDKException(ErrorCode.PARSE_ERROR, '没有可选项');
    }

    console.log('   选项:');
    options.forEach((opt, i) => {
      const label = this.getOptionLabel(opt, i);
      console.log(`   [${i + 1}] ${label}`);
    });

    const input = await this.prompt(`   请输入选项编号 (1-${options.length}): `);
    const selectedIndex = parseInt(input.trim(), 10);

    if (isNaN(selectedIndex) || selectedIndex < 1 || selectedIndex > options.length) {
      throw new SDKException(
        ErrorCode.PARSE_ERROR,
        `无效的选择: ${input}，请输入 1-${options.length} 之间的数字`
      );
    }

    const selectedOption = options[selectedIndex - 1];
    const decision = this.getOptionValue(selectedOption);

    return {
      callId,
      type: InteractionType.USER_SELECT,
      response: {
        selectedIndex: selectedIndex - 1,
        selectedValue: selectedOption,
      },
      source: this.extractSource(payload),
      modifiedData: this.extractData(payload),
      decision,
    };
  }

  /** 处理用户输入（表单模式）/ Handle user input (form mode) */
  async handleUserInput(payload: Record<string, unknown>, callId: string): Promise<InteractiveResponse> {
    const title = this.getTitle(payload);
    const description = this.getDescription(payload);
    const source = this.extractSource(payload);

    const formSpec = this.extractFormSpec(payload);
    const elements = this.getFormElements(formSpec);
    const initialValues = this.getFormInitialValues(formSpec);

    console.log(`\n✏️  ${title}`);
    if (description) console.log(`    ${description}`);
    console.log(`    ${'-'.repeat(40)}`);

    const formData: Record<string, unknown> = {};

    for (const elem of elements) {
      const field = this.getFieldKey(elem);
      const label = this.getFieldLabel(elem, field);
      const widget = this.getFieldWidget(elem);
      const placeholder = this.getFieldPlaceholder(elem);
      const defaultValue = this.getInitialValue(initialValues, field);

      if (widget === 'radio' || widget === 'segmented') {
        const enumOpts = this.getFieldEnum(formSpec, field);
        if (enumOpts.length > 0) {
          console.log(`    ${label}:`);
          enumOpts.forEach((opt, i) => {
            const marker = defaultValue === opt ? '*' : ' ';
            console.log(`      [${i + 1}]${marker} ${opt}`);
          });
          let prompt = `    请选择 (1-${enumOpts.length})`;
          if (defaultValue) prompt += ` [默认: ${defaultValue}]`;
          prompt += ': ';
          const input = (await this.prompt(prompt)).trim();
          if (!input && defaultValue) {
            formData[field] = defaultValue;
          } else {
            const idx = parseInt(input, 10);
            if (!isNaN(idx) && idx >= 1 && idx <= enumOpts.length) {
              formData[field] = enumOpts[idx - 1];
            } else {
              formData[field] = defaultValue;
            }
          }
        }
      } else {
        let prompt = `    ${label}`;
        if (placeholder) prompt += ` (${placeholder})`;
        if (defaultValue) prompt += ` [默认: ${defaultValue}]`;
        prompt += ': ';
        const input = (await this.prompt(prompt)).trim();
        if (!input && defaultValue) {
          formData[field] = defaultValue;
        } else {
          formData[field] = input;
        }
      }
    }

    console.log(`    ${'-'.repeat(40)}`);

    return {
      callId,
      type: InteractionType.USER_INPUT,
      response: { value: formData },
      source,
      formData,
      decision: 'submit',
    };
  }

  /** 使用交互响应恢复对话 / Resume chat with interactive response */
  async *resumeChat(
    threadId: string,
    response: InteractiveResponse,
    baseVariables?: Record<string, unknown>
  ): AsyncIterable<ChatEvent> {
    if (!this.client) {
      yield {
        rawJson: '',
        statusCode: 0,
        isDone: false,
        error: new SDKException(ErrorCode.CLIENT_CREATE, '客户端未初始化'),
      };
      return;
    }

    if (!response) {
      yield {
        rawJson: '',
        statusCode: 0,
        isDone: false,
        error: new SDKException(ErrorCode.PARSE_ERROR, '交互响应为空'),
      };
      return;
    }

    const userInteractive: Record<string, unknown> = {
      callId: response.callId,
      source: response.source,
      decision: response.decision,
    };
    if (response.type === InteractionType.USER_INPUT) {
      userInteractive.formData = response.formData;
    } else {
      userInteractive.modifiedData = response.modifiedData;
    }

    const uiJson = JSON.stringify(userInteractive);
    yield* this.client.interact(threadId, uiJson, baseVariables);
  }

  // =================================================================================
  // 辅助方法
  // =================================================================================

  private async prompt(question: string): Promise<string> {
    return new Promise((resolve) => {
      this.rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
      });

      this.rl.question(question, (answer) => {
        this.rl?.close();
        resolve(answer);
      });
    });
  }

  private getTitle(payload: Record<string, unknown>): string | undefined {
    const userAck = payload.userAck as Record<string, unknown> | undefined;
    if (userAck) {
      const data = userAck.data as Record<string, unknown> | undefined;
      if (data?.title) return data.title as string;
    }
    const userInput = payload.userInput as Record<string, unknown> | undefined;
    if (userInput?.title) return userInput.title as string;
    const meta = payload.meta as Record<string, unknown> | undefined;
    return meta?.title as string | undefined;
  }

  private getDescription(payload: Record<string, unknown>): string | undefined {
    const userAck = payload.userAck as Record<string, unknown> | undefined;
    if (userAck?.message) return userAck.message as string;
    const userInput = payload.userInput as Record<string, unknown> | undefined;
    if (userInput?.description) return userInput.description as string;
    const meta = payload.meta as Record<string, unknown> | undefined;
    return (meta?.description || meta?.desc) as string | undefined;
  }

  private getOptions(payload: Record<string, unknown>): Array<Record<string, unknown>> {
    const userAck = payload.userAck as Record<string, unknown> | undefined;
    if (userAck?.options) {
      const opts = userAck.options as Array<Record<string, unknown>>;
      if (opts.length > 0) return opts;
    }
    const data = payload.data as Array<Record<string, unknown>> | undefined;
    if (data && data.length > 0) return data;
    const meta = payload.meta as Record<string, unknown> | undefined;
    return (meta?.options as Array<Record<string, unknown>>) || [];
  }

  private getOptionLabel(option: Record<string, unknown>, index: number): string {
    for (const field of ['label', 'name', 'title', 'value']) {
      const value = option[field];
      if (typeof value === 'string' && value) return value;
    }
    return `选项 ${index + 1}`;
  }

  private getOptionValue(option: Record<string, unknown>): string {
    const value = option.value;
    return typeof value === 'string' ? value : '';
  }

  private extractSource(payload: Record<string, unknown>): Record<string, unknown> | undefined {
    const userAck = payload.userAck as Record<string, unknown> | undefined;
    if (userAck?.source) return userAck.source as Record<string, unknown>;
    const userInput = payload.userInput as Record<string, unknown> | undefined;
    if (userInput?.source) return userInput.source as Record<string, unknown>;
    const meta = payload.meta as Record<string, unknown> | undefined;
    return meta?.source as Record<string, unknown> | undefined;
  }

  private extractData(payload: Record<string, unknown>): Record<string, unknown> | undefined {
    const userAck = payload.userAck as Record<string, unknown> | undefined;
    if (userAck?.data) return userAck.data as Record<string, unknown>;
    const meta = payload.meta as Record<string, unknown> | undefined;
    return meta?.data as Record<string, unknown> | undefined;
  }

  // =================================================================================
  // formSpec 辅助方法 (user_input 表单模式)
  // =================================================================================

  private extractFormSpec(payload: Record<string, unknown>): Record<string, unknown> | undefined {
    const userInput = payload.userInput as Record<string, unknown> | undefined;
    return userInput?.formSpec as Record<string, unknown> | undefined;
  }

  private getFormElements(formSpec?: Record<string, unknown>): Array<Record<string, unknown>> {
    if (!formSpec) return [];
    const uiSchema = formSpec.ui_schema as Record<string, unknown> | undefined;
    return (uiSchema?.elements as Array<Record<string, unknown>>) || [];
  }

  private getFormInitialValues(formSpec?: Record<string, unknown>): Record<string, unknown> | undefined {
    return formSpec?.initialValues as Record<string, unknown> | undefined;
  }

  private getFieldKey(elem: Record<string, unknown>): string {
    return (elem.field as string) || '';
  }

  private getFieldLabel(elem: Record<string, unknown>, field: string): string {
    return (elem.label as string) || field;
  }

  private getFieldWidget(elem: Record<string, unknown>): string {
    return (elem.widget as string) || 'input';
  }

  private getFieldPlaceholder(elem: Record<string, unknown>): string {
    return (elem.placeholder as string) || '';
  }

  private getInitialValue(initialValues: Record<string, unknown> | undefined, field: string): string {
    if (!initialValues) return '';
    const val = initialValues[field];
    return val !== undefined ? String(val) : '';
  }

  private getFieldEnum(formSpec: Record<string, unknown> | undefined, field: string): string[] {
    if (!formSpec) return [];
    const schema = formSpec.schema as Record<string, unknown> | undefined;
    const properties = schema?.properties as Record<string, unknown> | undefined;
    const prop = properties?.[field] as Record<string, unknown> | undefined;
    const enumVals = prop?.enum as unknown[] | undefined;
    return (enumVals || []).map(v => String(v));
  }

  /** 检查事件是否为交互事件 / Check if event is interactive */
  static isInteractiveEvent(event: Record<string, unknown>): boolean {
    return event?.type === EventType.INTERACTIVE;
  }

  /** 从消息中提取交互事件 / Extract interactive events from message */
  static extractInteractiveEvents(
    message: Record<string, unknown>
  ): Array<Record<string, unknown>> {
    const events = message?.events as Array<Record<string, unknown>> | undefined;
    if (!events) return [];
    return events.filter((e) => InteractiveHandler.isInteractiveEvent(e));
  }
}

File: samples/python/starops_sdk_samples/client/interactive_handler.py
"""
Interactive handler for STAROps SDK
STAROps SDK 交互处理器
"""

import json
import sys
import time
from dataclasses import dataclass
from typing import Any, AsyncIterator, Dict, List, Optional, TextIO

from ..types import EventType, InteractionType
from .agent_client import AgentClient, ChatEvent
from .errors import SDKException, ErrorCode


@dataclass
class InteractiveResponse:
    """交互响应 / Interactive response"""
    call_id: str
    type: InteractionType
    response: Dict[str, Any]
    source: Optional[Dict[str, Any]] = None
    modified_data: Optional[Dict[str, Any]] = None
    form_data: Optional[Dict[str, Any]] = None
    decision: str = ""


class InteractiveHandler:
    """交互事件处理器 / Interactive event handler"""

    def __init__(
        self,
        client: AgentClient,
        timeout: Optional[float] = None,
        reader: Optional[TextIO] = None,
        writer: Optional[TextIO] = None,
    ):
        self.client = client
        self.timeout = timeout
        self.reader = reader or sys.stdin
        self.writer = writer or sys.stdout

    def handle_event(self, event: Dict[str, Any], call_id: str) -> InteractiveResponse:
        """处理交互事件 / Handle interactive event"""
        if not event:
            raise SDKException(ErrorCode.PARSE_ERROR, "事件为空")

        event_type = event.get("type")
        if event_type != EventType.INTERACTIVE.value:
            raise SDKException(ErrorCode.PARSE_ERROR, f"不支持的事件类型: {event_type}")

        payload = event.get("payload")
        if not payload:
            raise SDKException(ErrorCode.PARSE_ERROR, "交互负载为空")

        interactive_type = payload.get("type")

        if interactive_type == InteractionType.USER_ACK.value:
            return self.handle_user_ack(payload, call_id)
        elif interactive_type == InteractionType.USER_SELECT.value:
            return self.handle_user_select(payload, call_id)
        elif interactive_type == InteractionType.USER_INPUT.value:
            return self.handle_user_input(payload, call_id)
        else:
            raise SDKException(ErrorCode.PARSE_ERROR, f"不支持的交互类型: {interactive_type}")

    def handle_user_ack(self, payload: Dict[str, Any], call_id: str) -> InteractiveResponse:
        """处理用户确认 / Handle user acknowledgment"""
        title = self._get_title(payload)
        message = self._get_description(payload)
        options = self._get_options(payload)
        modified_data = self._extract_data(payload)

        self._print("\n🔔 确认请求")
        self._print("--------------")
        if title:
            self._print(title)
        if message:
            self._print(f"\n{message}")

        if modified_data:
            self._print("")
            for key, value in modified_data.items():
                if key in ("title", "message"):
                    continue
                self._print(f"{key}: {value}")
        self._print("--------------")

        if options:
            parts = []
            for i, opt in enumerate(options):
                parts.append(f"[{self._get_option_value(opt)}] {self._get_option_label(opt, i)}")
            self._print(f"请输入 {', '.join(parts)}: ", end="")
        else:
            self._print("请输入 [y/yes] 确认，[n/no] 取消: ", end="")

        user_input = self._read_input().strip().lower()
        confirmed = user_input in ("", "y", "yes", "是")
        decision = "yes" if confirmed else "no"

        for opt in options:
            if user_input == self._get_option_value(opt).lower():
                decision = self._get_option_value(opt)
                confirmed = True
                break

        return InteractiveResponse(
            call_id=call_id,
            type=InteractionType.USER_ACK,
            response={"confirmed": confirmed},
            source=self._extract_source(payload),
            modified_data=modified_data,
            decision=decision,
        )

    def handle_user_select(self, payload: Dict[str, Any], call_id: str) -> InteractiveResponse:
        """处理用户选择 / Handle user selection"""
        title = self._get_title(payload)

        self._print("\n📋 请选择")
        if title:
            self._print(f"   标题: {title}")

        options = self._get_options(payload)
        if not options:
            raise SDKException(ErrorCode.PARSE_ERROR, "没有可选项")

        self._print("   选项:")
        for i, opt in enumerate(options):
            label = self._get_option_label(opt, i)
            self._print(f"   [{i + 1}] {label}")
        self._print(f"   请输入选项编号 (1-{len(options)}): ", end="")

        user_input = self._read_input().strip()
        try:
            selected_index = int(user_input)
        except ValueError:
            raise SDKException(
                ErrorCode.PARSE_ERROR,
                f"无效的选择: {user_input}，请输入 1-{len(options)} 之间的数字"
            )

        if selected_index < 1 or selected_index > len(options):
            raise SDKException(
                ErrorCode.PARSE_ERROR,
                f"无效的选择: {user_input}，请输入 1-{len(options)} 之间的数字"
            )

        selected_option = options[selected_index - 1]
        decision = self._get_option_value(selected_option)

        return InteractiveResponse(
            call_id=call_id,
            type=InteractionType.USER_SELECT,
            response={
                "selectedIndex": selected_index - 1,
                "selectedValue": selected_option,
            },
            source=self._extract_source(payload),
            modified_data=self._extract_data(payload),
            decision=decision,
        )

    def handle_user_input(self, payload: Dict[str, Any], call_id: str) -> InteractiveResponse:
        """处理用户输入（表单模式）/ Handle user input (form mode)"""
        title = self._get_title(payload)
        description = self._get_description(payload)
        source = self._extract_source(payload)

        form_spec = self._extract_form_spec(payload)
        elements = self._get_form_elements(form_spec)
        initial_values = self._get_form_initial_values(form_spec)

        self._print(f"\n✏️  {title}")
        if description:
            self._print(f"    {description}")
        self._print(f"    {'-' * 40}")

        form_data = {}
        for elem in elements:
            field = self._get_field_key(elem)
            label = self._get_field_label(elem, field)
            widget = self._get_field_widget(elem)
            placeholder = self._get_field_placeholder(elem)
            default_value = self._get_initial_value(initial_values, field)

            if widget in ("radio", "segmented"):
                enum_opts = self._get_field_enum(form_spec, field)
                if enum_opts:
                    self._print(f"    {label}:")
                    for i, opt in enumerate(enum_opts):
                        marker = "*" if default_value == opt else " "
                        self._print(f"      [{i + 1}]{marker} {opt}")
                    prompt = f"    请选择 (1-{len(enum_opts)})"
                    if default_value:
                        prompt += f" [默认: {default_value}]"
                    prompt += ": "
                    user_input = self._do_prompt(prompt).strip()
                    if not user_input and default_value:
                        form_data[field] = default_value
                    else:
                        try:
                            idx = int(user_input)
                            if 1 <= idx <= len(enum_opts):
                                form_data[field] = enum_opts[idx - 1]
                            else:
                                form_data[field] = default_value
                        except ValueError:
                            form_data[field] = default_value
            else:
                prompt = f"    {label}"
                if placeholder:
                    prompt += f" ({placeholder})"
                if default_value:
                    prompt += f" [默认: {default_value}]"
                prompt += ": "
                user_input = self._do_prompt(prompt).strip()
                if not user_input and default_value:
                    form_data[field] = default_value
                else:
                    form_data[field] = user_input

        self._print(f"    {'-' * 40}")

        return InteractiveResponse(
            call_id=call_id,
            type=InteractionType.USER_INPUT,
            response={"value": form_data},
            source=source,
            form_data=form_data,
            decision="submit",
        )

    async def resume_chat(
        self, thread_id: str, response: InteractiveResponse,
        base_variables: Dict[str, Any] = None
    ) -> AsyncIterator[ChatEvent]:
        """使用交互响应恢复对话 / Resume chat with interactive response"""
        if not self.client:
            yield ChatEvent.from_error(
                SDKException(ErrorCode.CLIENT_CREATE, "客户端未初始化")
            )
            return

        if not response:
            yield ChatEvent.from_error(
                SDKException(ErrorCode.PARSE_ERROR, "交互响应为空")
            )
            return

        user_interactive = {
            "callId": response.call_id,
            "source": response.source,
            "decision": response.decision,
        }
        if response.type == InteractionType.USER_INPUT:
            user_interactive["formData"] = response.form_data
        else:
            user_interactive["modifiedData"] = response.modified_data

        ui_json = json.dumps(user_interactive)
        async for event in self.client.interact(thread_id, ui_json, base_variables):
            yield event

    # =================================================================================
    # 辅助方法
    # =================================================================================

    def _print(self, text: str, end: str = "\n") -> None:
        self.writer.write(text + end)
        self.writer.flush()

    def _do_prompt(self, prompt: str) -> str:
        self.writer.write(prompt)
        self.writer.flush()
        try:
            return self.reader.readline()
        except Exception as e:
            raise SDKException(ErrorCode.PARSE_ERROR, "读取输入失败", e)

    def _read_input(self) -> str:
        try:
            return self.reader.readline()
        except Exception as e:
            raise SDKException(ErrorCode.PARSE_ERROR, "读取输入失败", e)

    def _get_title(self, payload: Dict[str, Any]) -> Optional[str]:
        user_ack = payload.get("userAck", {})
        if user_ack:
            data = user_ack.get("data", {})
            if data.get("title"):
                return data["title"]
        user_input = payload.get("userInput", {})
        if user_input.get("title"):
            return user_input["title"]
        meta = payload.get("meta", {})
        return meta.get("title")

    def _get_description(self, payload: Dict[str, Any]) -> Optional[str]:
        user_ack = payload.get("userAck", {})
        if user_ack.get("message"):
            return user_ack["message"]
        user_input = payload.get("userInput", {})
        if user_input.get("description"):
            return user_input["description"]
        meta = payload.get("meta", {})
        return meta.get("description") or meta.get("desc")

    def _get_options(self, payload: Dict[str, Any]) -> List[Dict[str, Any]]:
        user_ack = payload.get("userAck", {})
        options = user_ack.get("options")
        if options:
            return options
        data = payload.get("data", [])
        if data:
            return data
        meta = payload.get("meta", {})
        return meta.get("options", [])

    def _get_option_label(self, option: Dict[str, Any], index: int) -> str:
        for field in ["label", "name", "title", "value"]:
            value = option.get(field)
            if isinstance(value, str) and value:
                return value
        return f"选项 {index + 1}"

    def _get_option_value(self, option: Dict[str, Any]) -> str:
        value = option.get("value")
        return str(value) if value is not None else ""

    def _extract_source(self, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        user_ack = payload.get("userAck", {})
        if user_ack.get("source"):
            return user_ack["source"]
        user_input = payload.get("userInput", {})
        if user_input.get("source"):
            return user_input["source"]
        meta = payload.get("meta", {})
        return meta.get("source")

    def _extract_data(self, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        user_ack = payload.get("userAck", {})
        if user_ack.get("data"):
            return user_ack["data"]
        meta = payload.get("meta", {})
        return meta.get("data")

    # =================================================================================
    # formSpec 辅助方法 (user_input 表单模式)
    # =================================================================================

    def _extract_form_spec(self, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        user_input = payload.get("userInput", {})
        return user_input.get("formSpec")

    def _get_form_elements(self, form_spec: Optional[Dict[str, Any]]) -> List[Dict[str, Any]]:
        if not form_spec:
            return []
        ui_schema = form_spec.get("ui_schema", {})
        return ui_schema.get("elements", [])

    def _get_form_initial_values(self, form_spec: Optional[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
        if not form_spec:
            return None
        return form_spec.get("initialValues")

    def _get_field_key(self, elem: Dict[str, Any]) -> str:
        return elem.get("field", "")

    def _get_field_label(self, elem: Dict[str, Any], field: str) -> str:
        return elem.get("label", field)

    def _get_field_widget(self, elem: Dict[str, Any]) -> str:
        return elem.get("widget", "input")

    def _get_field_placeholder(self, elem: Dict[str, Any]) -> str:
        return elem.get("placeholder", "")

    def _get_initial_value(self, initial_values: Optional[Dict[str, Any]], field: str) -> str:
        if not initial_values:
            return ""
        val = initial_values.get(field)
        return str(val) if val is not None else ""

    def _get_field_enum(self, form_spec: Optional[Dict[str, Any]], field: str) -> List[str]:
        if not form_spec:
            return []
        schema = form_spec.get("schema", {})
        properties = schema.get("properties", {})
        prop = properties.get(field, {})
        enum_vals = prop.get("enum", [])
        return [str(v) for v in enum_vals]

    @staticmethod
    def is_interactive_event(event: Dict[str, Any]) -> bool:
        """检查事件是否为交互事件 / Check if event is interactive"""
        if not event:
            return False
        return event.get("type") == EventType.INTERACTIVE.value

    @staticmethod
    def extract_interactive_events(message: Dict[str, Any]) -> List[Dict[str, Any]]:
        """从消息中提取交互事件 / Extract interactive events from message"""
        events = []
        if not message:
            return events

        msg_events = message.get("events", [])
        for event in msg_events:
            if InteractiveHandler.is_interactive_event(event):
                events.append(event)
        return events